projects/telemetry/src/nmea_checksum.c

```diff
@@ -0,0 +1,79 @@
+#include "nmea_checksum.h"
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+// Private method to convert hex char to int
+static uint8_t prv_hex_to_int(char h) {
+  if ('0' <= h && h <= '9') {
+    return h - '0';
+  } else if ('A' <= h && h <= 'F') {
+    // We only care about uppercase because the NMEA messages are all caps
+    return h - 'A' + 10;
+  }
+  // Just to make the checksum fail
+  return 16;
+}
+
+static StatusCode prv_checksum_to_int(char tens, char ones, uint8_t *computed) {
+  uint8_t int_tens = prv_hex_to_int(tens);
+  uint8_t int_ones = prv_hex_to_int(ones);
+
+  // Checks for the 16 because it is the result sent in case of error
+  if (ones == 16 || tens == 16) {
+    return STATUS_CODE_INVALID_ARGS;
+  }
+  if (computed != NULL) {
+    *computed = int_tens * 16 + int_ones;
+  }
+  return STATUS_CODE_OK;
+}
+
+// The checksum is calculated by taking the XOR of all characters between
+// (but not including) the '$' and '*' character
+// An example NMEA message is "$GPGLL,2503.6319,N,12136.0099,E,053740.000,A,A*52"
+// In the message above, the checksum would be 52.
+StatusCode nmea_checksum_compute(char *message, size_t message_len, uint8_t *checksum) {
+  if (checksum == NULL || message == NULL || message_len < 4) {
+    return status_msg(STATUS_CODE_INVALID_ARGS,
+                      "out or message cannot be NULL. message_len must be greater than 3");
+  }
+
+  uint8_t sum = 0;
+  // start at the 2nd character, since the first character is $
+  for (uint8_t i = 1; message[i] != '*' && i < (message_len - 1); i++) {
+    sum ^= message[i];
+  }
+  *checksum = sum;
+  return STATUS_CODE_OK;
+}
+
+bool nmea_checksum_validate(char *message, size_t message_len) {
+  if (message == NULL || message_len < 4 || message[message_len - 1] != '\0') {
+    return false;
+  }
+
+  // We expect the last 2 characters in the message to be the sent checksum
+  // Hence, the 3rd last character should be '*' (we subtract 3 instead of 2 to
+  // account for \0)
+  char *received = message + message_len - 3;
```

Comment:
You might as well just treat it as a char array and index into message instead of manual pointer math.